backend/app/controllers.py

```diff
@@ -0,0 +1,234 @@
+        """Returns an empty response.
+
+        Used for API health check endpoint.
+
+        :return: An empty response.
+        :rtype: dict
+        """
+        return responses.JSONResponse({})
+
+
+class Login(Base):
+    """Auth
+
+    Handles Login API requests
+    """
+
+    def __init__(self, app: bottle.Bottle = None):
+        super().__init__(app)
+        self._request = services.Request('login')
+        self._login_authorization_uri = f'{settings.LOGIN_APP_URL}' \
+            f'?grant_type={settings.GRANT_TYPES["authorization"]}' \
+            f'&response_type={settings.RESPONSE_TYPE}' \
+            f'&redirect_uri={settings.REDIRECT_URL}' \
+            f'&client_id={settings.CLIENT_ID}'
+
+    def exchange_token(self) -> responses.JSONResponse:
+        """ Pass request to login app
+
+        :return: Access token data.
+        :rtype: responses.JSONResponse
+        """
+
+        code = bottle.request.params.get('code', None)
+
+        if code is None:
+            return responses.JSONResponse(
+                body={'message': 'Authorization failed'},
+                status=406
+            )
+
+        auth_api_response = self._request.post('/exchangeToken', {
+            'client_secret': settings.CLIENT_SECRET,
+            'client_id': settings.CLIENT_ID,
+            'redirect_uri': settings.REDIRECT_URL,
+            'grant_type': settings.GRANT_TYPES['exchange_token'],
+            'code': code
+        })
+
+        if auth_api_response.ok:
+            data = auth_api_response.json()
+
+            response = responses.JSONResponse(
+                headers={
+                    'Location': settings.APP_URL
+                },
+                status=303)
+
+            # https://github.com/bottlepy/bottle/blob/master/bottle.py#L1848
+            # In bottle starting from 0.13-dev
+            Morsel._reserved.setdefault('samesite', 'SameSite')
+
+            response.set_cookie('Authorization',
+                                f'Bearer {data["access_token"]}',
+                                max_age=data['expires_in'],
+                                httponly=True,
+                                secure=settings.SSL_ENABLED,
+                                samesite='Strict')
+
+            return response
+        else:
+            return responses.JSONResponse(body=auth_api_response.text,
+                                          status=auth_api_response.status_code)
+
+    def me(self) -> responses.JSONResponse:
+        """Pass request to login app
+
+        :return: Access token data.
+        :rtype: responses.JSONResponse
+        """
+
+        response = self._request.get(
+            '/me',
+            authorization_token=bottle.request.get_cookie(
+                'Authorization')
+        )
+
+        return responses.JSONResponse(body=response.text,
+                                      status=response.status_code)
+
+    def login(self) -> responses.JSONResponse:
```

Comment:
This won't do.

There should be no API endpoint where the frontend requests an URL from.

There should be an endpoint where the frontend redirects the user to, and it generates the authorization URI, *with a state* and then redirects the user there.

Reply:
Should we include state validation code in sample app?

Reply:
Yes, it should be a sample of how to do things correctly, and including a state is a good practice so we should demonstrate it.

You can use some simple `generate_state` and `validate_state` -methods which e.g. simply use a HMAC hash of the current timestamp and that the validated timestamp is at most an hour old or some such and that should give sufficient guarantees that the state validation is good enough.